climada/engine/unsequa/calc_delta_climate.py

```diff
@@ -0,0 +1,435 @@
+
+You should have received a copy of the GNU General Public License along
+with CLIMADA. If not, see <https://www.gnu.org/licenses/>.
+
+---
+
+Define Uncertainty class for Impact differences between two climates.
+"""
+
+__all__ = ["CalcDeltaImpact"]
+
+import logging
+import time
+from typing import Union
+import itertools
+
+import pandas as pd
+import numpy as np
+import pathos.multiprocessing as mp
+
+# use pathos.multiprocess fork of multiprocessing for compatibility
+# wiht notebooks and other environments https://stackoverflow.com/a/65001152/12454103
+
+from climada.engine import ImpactCalc
+from climada.engine.unsequa import Calc, InputVar, UncImpactOutput
+from climada.engine.unsequa.calc_base import (
+    _sample_parallel_iterator,
+    _multiprocess_chunksize,
+    _transpose_chunked_data,
+)
+from climada.entity import Exposures, ImpactFuncSet
+from climada.hazard import Hazard
+from climada.util import log_level
+from climada.util.value_representation import safe_divide
+
+
+LOGGER = logging.getLogger(__name__)
+
+
+class CalcDeltaImpact(Calc):
+    """
+    Delta Impact uncertainty caclulation class.
+
+    This is the class to perform uncertainty analysis on the outputs of a relative
+    change in impact from a (final impact - initial impact) / initial impact.
+    Impact objects are regular climada.engine.impact.Impact() objects. The resulting
+    Delta Impact is a relative change (fraction of the inital impact). The relative
+    change is intuitive to understand in contrast to absolute changes which are
+    hard to understand without knowledge of the absolute initial (baseline) state.
+
+    Attributes
+    ----------
+    rp : list(int)
+        List of the chosen return periods.
+    calc_eai_exp : bool
+        Compute eai_exp or not
+    calc_at_event : bool
+        Compute eai_exp or not
+    value_unit : str
+        Unit of the exposures value
+    exp_input_var : InputVar or Exposures
+        Exposure uncertainty variable
+    impf_input_var : InputVar if ImpactFuncSet
+        Impact function set uncertainty variable
+    haz_input_var: InputVar or Hazard
+        Hazard uncertainty variable
+    _input_var_names : tuple(str)
+        Names of the required uncertainty input variables
+        ('exp_initial_input_var', 'impf_initial_input_var', 'haz_initial_input_var',
+         'exp_final_input_var', 'impf_final_input_var', 'haz_final_input_var'')
+    _metric_names : tuple(str)
+        Names of the impact output metrics
+        ('aai_agg', 'freq_curve', 'at_event', 'eai_exp')
+    """
+
+    _input_var_names = (
+        "exp_initial_input_var",
+        "impf_initial_input_var",
+        "haz_initial_input_var",
+        "exp_final_input_var",
+        "impf_final_input_var",
+        "haz_final_input_var",
+    )
+    """Names of the required uncertainty variables"""
+
+    _metric_names = ("aai_agg", "freq_curve", "at_event", "eai_exp")
+    """Names of the cost benefit output metrics"""
```

Comment:
Why do these attributes exist? If I am not mistaken, they are never used anywhere in this class.

Reply:
Ah, I see they are used in the base class. This is super confusing! In particular, the empty attributes of `Calc` with the same name are never overwritten here. @chahank, instead of relying of `self` to point to the derived class, it would probably be better to define `Calc.input_vars` as `@abstractmethod`, to clarify what needs to be overwritten by derived classes. See https://docs.python.org/3/library/abc.html#abc.abstractmethod

Reply:
Alternative: Make `input_var_names` and `metric_names` parameters of the `Calc` base class and pass them through the base class constructor? This way, it would also be more explicit.

Reply:
@peanutfun : thank you for the insights on the code struture! All elements that one could consider, but I do not think that they are critical. I also think this PR is not the place for such discussion. The module is at the moment stable and functional. I would suggest to have a discussion on the needs of refactoring somewhere else.

Reply:
Absolutely agreed! My comments were a rather unordered stream of thoughts 🙈 I'll try to keep track of that in a separate Discussion, and resolve this conversation!